pkg/kubelet/cm/device_plugin_handler.go

```diff
@@ -205,21 +362,102 @@ func (h *DevicePluginHandlerImpl) Allocate(pod *v1.Pod, container *v1.Container,
 		// requests may fail if we serve them in mixed order.
 		// TODO: may revisit this part later if we see inefficient resource allocation
 		// in real use as the result of this.
-		resp, err := h.devicePluginManager.Allocate(resource, append(devices.UnsortedList(), allocated...))
+		resp, err := h.devicePluginManager.Allocate(resource, allocDevices)
 		if err != nil {
-			return nil, err
+			return fmt.Errorf("requested number of devices unavailable for %s. Requested: %d", resource, needed)
 		}
-		ret = append(ret, resp)
+		// Update internal cached podDevices state.
+		h.Lock()
+		allocDeviceSet := sets.NewString(allocDevices...)
+		if oldDevices := h.podDevices.containerDevices(string(pod.UID), container.Name, resource); oldDevices != nil {
+			freeDevices := oldDevices.Difference(allocDeviceSet)
+			h.allocatedDevices[resource] = h.allocatedDevices[resource].Difference(freeDevices)
+		}
+		h.podDevices.insert(string(pod.UID), container.Name, resource, allocDeviceSet, *resp)
+		h.Unlock()
 	}
+
 	// Checkpoints device to container allocation information.
 	if err := h.writeCheckpoint(); err != nil {
-		return nil, err
+		return err
 	}
-	return ret, nil
+	return nil
 }
 
-// updateAllocatedDevices updates the list of GPUs in use.
-// It gets a list of active pods and then frees any GPUs that are bound to
+func (h *DevicePluginHandlerImpl) Allocate(node *schedulercache.NodeInfo, attrs *lifecycle.PodAdmitAttributes) error {
+	pod := attrs.Pod
+	pods := attrs.OtherPods
+	// TODO: determine whether to take initContainer into consideration
+	for _, container := range pod.Spec.Containers {
+		if devOpts := h.GetDeviceRunContainerOptions(pod, &container); devOpts != nil {
+			continue
+		}
+		// No runtime information found in cache, make allocation
+		if err := h.allocateContainerResources(pod, &container); err != nil {
+			return err
+		}
+
+	}
+	// Update nodeInfo, to ensure NodeInfo.allocatableResource
```

Comment:
Suggest we add a private function, say sanitizeNodeAllocatable(), in DevicePluginHandlerImpl to update nodeInfo.allocatableResource and add a unit test for it. I am still not quite sure the need to scan through other pods. As I understand, as long as we set nodeInfo.AllocatableResource to be at least of pod's required resource, the pod can pass GeneralPredicate. Or did I miss something here? Do you think whether we may simplify allocatableResource update logic by using DevicePluginHandlerImpl.allocatedDevices(). I.e., something like below?

if !podRequireDevicePluginResource {
  return nil
}
var newAllocatableResource *Resouce
currentAllocatable := node.AllocatableResource()
for resource, quant := currentAllocatable.ScalarResources() {
  if !h.isDevicePluginResource(resource) {
    continue
  }
  needed := len(h.allocatedDevices[resource])
  if quant <= needed {
    continue
  }
  // Needs to update nodeInfo.AllocatableResource to make sure ...
  if newAllocatableResource == nil {
    newAllocatableResource = currentAllocatable.Clone()
  }
  newAllocatableResource.ScalarResources[resource] = needed
}
if newAllocatableResource != nil {
  node.SetAllocatableResource(newAllocatableResource)
}

Reply:
Yes, I fall into the trap, ignoring that requested resources of existing pods could be fetched from _h.allocatedDevices_.
What I'm concerned about is how would it behavior in node capacity when plugin fails after kubelet restart. Does it exists in node.AllocatableResource with capacity 0 or does not exist at all?  If the latter, I think we could not get the type of requested resources from  node.AllocatableResource, and the scan through the new pod may still be needed.

Reply:
That is a good point! The current plan is to simply remove capacity from node status for deleted or failed device plugins. However, h.allocatedDevices should still contain in-use devices even for deleted resources. So maybe instead of scan through node's AllocatableResouce(), we may scan through h.allocatedDevices() like this:

if !podRequireDevicePluginResource {
  return nil
}
var newAllocatableResource *Resouce
allocatableScalarResource := node.AllocatableResource().ScalarResources() {
for resource, devices := range h.allocatedDevices {
  needed := devices.Len()
  quant, ok := allocatableScalarResource[resource]
  if ok && quant <= needed {
    continue
  }
  // Needs to update nodeInfo.AllocatableResource to make sure ...
  if newAllocatableResource == nil {
    newAllocatableResource = currentAllocatable.Clone()
  }
  newAllocatableResource.ScalarResources[resource] = needed
}
if newAllocatableResource != nil {
  node.SetAllocatableResource(newAllocatableResource)
}